File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/content/page/region/FragmentComponent.ts
module api.content.page.region {

    import PropertyTree = api.data.PropertyTree;
    import PropertyEvent = api.data.PropertyEvent;

    export class FragmentComponent extends Component implements api.Equitable, api.Cloneable {

        public static PROPERTY_FRAGMENT = 'fragment';

        public static PROPERTY_CONFIG = 'config';

        public static debug: boolean = false;

        private disableEventForwarding: boolean;

        private fragment: api.content.ContentId;

        private config: PropertyTree;

        private configChangedHandler: (event: PropertyEvent) => void;

        constructor(builder: FragmentComponentBuilder) {
            super(builder);

            this.fragment = builder.fragment;
            this.config = builder.config;
            this.configChangedHandler = (event: PropertyEvent) => {
                if (FragmentComponent.debug) {
                    console.debug("FragmentComponent[" + this.getPath().toString() + "].config.onChanged: ", event);
                }
                if (!this.disableEventForwarding) {
                    this.notifyPropertyValueChanged(FragmentComponent.PROPERTY_CONFIG);
                }
            };

            this.config.onChanged(this.configChangedHandler);
        }

        setDisableEventForwarding(value: boolean) {
            this.disableEventForwarding = value;
        }

        getFragment(): api.content.ContentId {
            return this.fragment;
        }

        getConfig(): PropertyTree {
            return this.config;
        }

        setFragment(contentId: api.content.ContentId, name: string) {
            var oldValue = this.fragment;
            this.fragment = contentId;

            this.setName(name ? new ComponentName(name) : this.getType().getDefaultName());

            if (!api.ObjectHelper.equals(oldValue, contentId)) {
                this.notifyPropertyChanged(FragmentComponent.PROPERTY_FRAGMENT);
            }
        }

        hasFragment(): boolean {
            return !!this.fragment;
        }

        doReset() {
            this.setFragment(null, null);
        }

        isEmpty(): boolean {
            return !this.fragment;
        }

        toJson(): ComponentTypeWrapperJson {

            var json: FragmentComponentJson = <FragmentComponentJson>super.toComponentJson();
            json.fragment = this.fragment != null ? this.fragment.toString() : null;
            json.config = this.config != null ? this.config.toJson() : null;

            return <ComponentTypeWrapperJson> {
                FragmentComponent: json
            };
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, FragmentComponent)) {
                return false;
            }

            var other = <FragmentComponent>o;

            if (!super.equals(o)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.fragment, other.fragment)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.config, other.config)) {
                return false;
            }

            return true;
        }

        clone(): FragmentComponent {
            return new FragmentComponentBuilder(this).build();
        }
    }

    export class FragmentComponentBuilder extends ComponentBuilder<FragmentComponent> {

        fragment: api.content.ContentId;

        config: PropertyTree;

        constructor(source?: FragmentComponent) {
            super(source);
            if (source) {
                this.fragment = source.getFragment();
                this.config = source.getConfig() ? source.getConfig().copy() : null;
            } else {
                this.config = new PropertyTree();
            }
            this.setType(FragmentComponentType.get());
        }

        public setFragment(value: api.content.ContentId): FragmentComponentBuilder {
            this.fragment = value;
            return this;
        }

        public setConfig(value: PropertyTree): FragmentComponentBuilder {
            this.config = value;
            return this;
        }

        public fromJson(json: FragmentComponentJson, region: Region): FragmentComponentBuilder {

            if (json.fragment) {
                this.setFragment(new api.content.ContentId(json.fragment));
            }

            this.setName(json.name ? new ComponentName(json.name) : null);


            if (json.config) {
                this.setConfig(PropertyTree.fromJson(json.config));
            }

            this.setParent(region);

            return this;
        }

        public build(): FragmentComponent {
            return new FragmentComponent(this);
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/region/FragmentComponent.java
package com.enonic.xp.region;

import java.util.Objects;

import com.enonic.xp.annotation.PublicApi;
import com.enonic.xp.content.ContentId;

@PublicApi
public final class FragmentComponent
    extends Component
{
    private final ContentId fragment;

    public FragmentComponent( final Builder builder )
    {
        super( builder );
        this.fragment = builder.fragment;
    }

    public static Builder create()
    {
        return new Builder();
    }

    public static Builder create( final FragmentComponent source )
    {
        return new Builder( source );
    }

    @Override
    public FragmentComponent copy()
    {
        return create( this ).build();
    }

    @Override
    public ComponentType getType()
    {
        return FragmentComponentType.INSTANCE;
    }

    public ContentId getFragment()
    {
        return fragment;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }
        if ( !super.equals( o ) )
        {
            return false;
        }
        final FragmentComponent that = (FragmentComponent) o;
        return Objects.equals( fragment, that.fragment );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( super.hashCode(), fragment );
    }

    public static final class Builder
        extends Component.Builder
    {
        private ContentId fragment;

        private Builder()
        {
            // Default
        }

        private Builder( final FragmentComponent source )
        {
            super( source );
            fragment = source.fragment;
        }

        public Builder fragment( final ContentId value )
        {
            this.fragment = value;
            return this;
        }

        public FragmentComponent build()
        {
            return new FragmentComponent( this );
        }
    }
}
